modules/data-streams/src/main/java/org/elasticsearch/datastreams/UpdateTimeSeriesRangeService.java

```diff
@@ -0,0 +1,191 @@
+                    LOGGER.warn("failed to update tsdb data stream end times", e);
+                    onComplete.run();
+                }
+
+            }, ClusterStateTaskExecutor.unbatched());
+        } else {
+            LOGGER.debug("not starting tsdb update task, because another execution is still running");
+        }
+    }
+
+    void setPollInterval(TimeValue newValue) {
+        LOGGER.info(
+            "updating [{}] setting from [{}] to [{}]",
+            DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.getKey(),
+            pollInterval,
+            newValue
+        );
+        this.pollInterval = newValue;
+    }
+
+    ClusterState updateTimeSeriesTemporalRange(ClusterState current, Instant now) {
+        Metadata.Builder mBuilder = null;
+        for (DataStream dataStream : current.metadata().dataStreams().values()) {
+            if (dataStream.getIndexMode() != IndexMode.TIME_SERIES) {
+                continue;
+            }
+            if (dataStream.isReplicated()) {
+                continue;
+            }
+
+            // getWriteIndex() selects the latest added index:
+            Index head = dataStream.getWriteIndex();
+            IndexMetadata im = current.metadata().getIndexSafe(head);
+            Instant currentEnd = IndexSettings.TIME_SERIES_END_TIME.get(im.getSettings());
+            TimeValue lookAheadTime = IndexSettings.LOOK_AHEAD_TIME.get(im.getSettings());
+            Instant newEnd = now.plus(lookAheadTime.getMillis(), ChronoUnit.MILLIS).plus(pollInterval.getMillis(), ChronoUnit.MILLIS);
+            if (newEnd.isAfter(currentEnd)) {
+                try {
+                    Settings settings = Settings.builder()
+                        .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), DEFAULT_DATE_TIME_FORMATTER.format(newEnd))
+                        .build();
+                    LOGGER.debug(
+                        "updating [{}] setting from [{}] to [{}] for data stream [{}]",
+                        IndexSettings.TIME_SERIES_END_TIME.getKey(),
+                        currentEnd,
+                        newEnd,
+                        dataStream.getName()
+                    );
+                    if (mBuilder == null) {
+                        mBuilder = Metadata.builder(current.metadata());
+                    }
+                    mBuilder.updateSettings(settings, head.getName());
+                    Metadata.Builder finalMBuilder = mBuilder;
+                    // Verify that all temporal ranges of each backing index is still valid:
+                    dataStream.validate(finalMBuilder::get);
+                } catch (Exception e) {
+                    LOGGER.error(
+                        "unable to update [{}] for data stream [{}] and backing index [{}]",
+                        IndexSettings.TIME_SERIES_END_TIME.getKey(),
+                        dataStream.getName(),
+                        head.getName()
+                    );
```

Comment:
I think this should include the exception that was thrown, or else it will be difficult to determine why it was unable to update the DS. If we're concerned about verbosity maybe we can make the inclusion of the cause predicated on the logging level?

Reply:
Oops yes, we should include the exception..

Reply:
Implemented via: f3aab54a6db450c668d6d7ef32e5823e7a1b0c0f